plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergBucketFunction.java

```diff
@@ -87,6 +131,11 @@ public int getBucket(Page page, int position)
         for (int i = 0; i < partitionColumns.size(); i++) {
             PartitionColumn partitionColumn = partitionColumns.get(i);
             Block block = page.getBlock(partitionColumn.getSourceChannel());
+            if (partitionColumn.getDepth().isPresent()) {
+                for (int index : partitionColumn.getDepth().get()) {
+                    block = block.getChildren().get(index);
```

Comment:
Why do we consider `getChildren()` to be unreliable here ?
It looks quite wasteful to do all the work of `toColumnarRow` for every single position.
If using `toColumnarRow` is the only way it works, then I would wonder implementing https://github.com/trinodb/trino/issues/9848 to batch this function.
cc: @findepi

Reply:
that's a good question. I am not sure when it's safe to use `block.getChildren()` vs `ColumnarRow`
I remember @djsstarburst fixing some correctness(?) issues in MERGE by switching from the former to the latter.

If this is needed, I can surely revive https://github.com/trinodb/trino/pull/12734. I stopped working on it because this is a performance-related stuff and I didn't have a clear performance-related objective to pursue. Sounds like now I have.

Reply:
I think we should not block the PR on bucket functions improvements like https://github.com/trinodb/trino/issues/9848, as we don't know how often bucketing on nested fields happens in practice. Let's have a correct implementation first, and maybe some TODO comment about future possible improvements

Reply:
`getChildren()` won't work here if `RowBlock` can have nulls (and I think it can). Then `RowBlock` position does not map directly to the child/field `Block` position because the position can be null directly in the main `RowBlock` or some previous position is null

Reply:
Thanks @lukasz-stec , so it looks like we must use `toColumnarRow`.
If this functionality was important enough to implement, I think we should ensure that the performance is not prohibitively bad. Could we do a benchmark comparing the CPU time taken for partitioning a table on a primitive field vs partitioning the same table on the same primitive field stored inside a ROW ? If the results of the benchmarks are in the same ballpark, then it would go okay to go ahead with current implementation.